frontend/components/tree/js/tree.service.ts

```diff
@@ -936,56 +993,105 @@ export class TreeService {
 	}
 
 	/**
 	 * Set selection status of node.
-	 * @param node		Node to set.
-	 * @param isSelected	Is selected.
+	 * @param node	Node to set selection status of
+	 * @param select whether the node should be selected or not
 	 */
-	public setNodeSelection(node: any, isSelected: boolean) {
-		if (node.selected !== isSelected) {
-			const nodeIndex = this.currentSelectedNodes.indexOf(node);
-
-			if (isSelected) {
-				if (-1 === nodeIndex) {
-					node.selected = true;
-					this.currentSelectedNodes.push(node);
-				}
-			} else {
-				if (nodeIndex > -1) {
-					this.currentSelectedNodes[nodeIndex].selected = false;
-					this.currentSelectedNodes.splice(nodeIndex, 1);
-				}
+	public setNodeSelection(node: any, select: boolean) {
+
+		// If node
+		if (node.selected === select) {
+			return;
+		}
+
+		const nodeIndex = this.currentSelectedNodes.indexOf(node);
+
+		if (select) {
+			if (nodeIndex === -1) {
+				node.selected = true;
+				this.currentSelectedNodes.push(node);
+			}
+		} else {
+			if (nodeIndex > -1) {
+				this.currentSelectedNodes[nodeIndex].selected = false;
+				this.currentSelectedNodes.splice(nodeIndex, 1);
 			}
+		}
+	}
 
-			return this.ready.promise.then(() => {
-				const currentSelectedMap = {};
-				this.currentSelectedNodes.forEach((n) => {
-					this.traverseNodeAndPushId(n, currentSelectedMap, this.treeMap.idToMeshes);
-				});
+	public getMeshHighlights(nodes) {
+		return this.ready.promise.then(() => {
+			const currentSelectedMap = this.getMeshMapFromNodes(nodes, this.treeMap.idToMeshes);
+			return currentSelectedMap;
+		});
+	}
 
-				this.setHighlightMap(currentSelectedMap);
-				return currentSelectedMap;
-			});
+	/**
+	 * Return a map of currently selected meshes
+	 */
+	public getCurrentMeshHighlights() {
+		return this.getMeshHighlights(this.currentSelectedNodes.concat());
+	}
+
+	/**
+	 * Deselect a nodes in the tree.
+	 * @param nodes	Node to select.
+	 */
+	public deselectNodes(nodes: any[]) {
+
+		for (let i = 0; i < nodes.length; i++) {
+			const node = nodes[i];
+			this.setNodeSelection(node, false);
 		}
+
+		return this.highlightNodes(nodes, true, undefined, false);
 	}
 
 	/**
-	 * Select a node in the tree.
-	 * @param node	Node to select.
+	 * Select nodes in the tree.
+	 * @param nodes	Nodes to select.
 	 * @param multi	Is multi select enabled.
 	 */
-	public selectNode(node: any, multi: boolean, final: boolean) {
+	public selectNodes(
+		nodes: any[], multi: boolean, additive: boolean, colour: number[],
+		forceReHighlight: boolean,
+	) {
+
+		forceReHighlight = (forceReHighlight === undefined) ? true : false;
+
+		if (!multi) {
+			// If it is not multiselect mode, remove all highlights
+			this.clearCurrentlySelected();
+		}
+
+		if (!nodes || nodes.length === 0) {
+			return Promise.resolve("No nodes specified");
+		}
+
+		for (let i = 0; i < nodes.length; i++) {
+			const node = nodes[i];
 
-		if (node) {
-			if (multi) {
+			if (!node) {
+				continue;
+			}
+
+			if (additive) {
```

Comment:
I'm pretty sure additive === forceHighlight. Please investigate!

Reply:
Actually this whole if block is probably a lot cleaner if we just do something like this:

````js
const shouldSelectNode = !multi || forceHighlight || !node.selected;
this.setNodeSelection(node, shouldSelectNode);
````

Currently we have 3 if statement blocks here 2 of which does something identical, the 3rd do something identical with a slight variation on the parameter

Reply:
Additive removed now as it was the same as forceHighlight. Not sure I am convinced about the one liner, you'd change the number of lines from 7 lines to 5 but I would say it makes it harder to follow as it's less explicit.

Reply:
From what I understand this should change from  7 lines to 2 lines? maybe you need to commit the change first and we can re-visit as i'm not really seeing it.

In general I'm pretty against the idea of having a multi conditional if statement that does the same thing but tweaking a single variable, because it gives the illusion that they're doing different things when they're not. At minimum the `setNodeSelection()` call should sit outside and the if statement should only define the variation of the 2nd parameter for the function call. but i'd argue that is no less clearer than a one liner.